File: apps/android/app/src/main/java/me/thanish/prayers/se/routes/RouteSpec.kt
package me.thanish.prayers.se.routes

import android.content.Context
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.navigation.NavController
import me.thanish.prayers.se.routes.compass.CompassRouteSpec
import me.thanish.prayers.se.routes.home.MainRouteSpec
import me.thanish.prayers.se.routes.settings.SettingsRouteSpec

/**
 * Describes the type of route.
 */
enum class RouteType {
    PRIMARY,  // Route is included in main navigation
    SECONDARY // Route is not included in main navigation
}

/**
 * Describes the structure of a route.
 */
data class RouteSpec(
    val name: String,
    val text: Int,
    val type: RouteType = RouteType.PRIMARY,
    val icon: @Composable () -> Pair<ImageVector, ImageVector>,
    val content: @Composable (NavController, Modifier) -> Unit
) {
    /**
     * getLabel returns the label for the route.
     */
    fun getLabel(context: Context): String {
        return context.getString(text)
    }
}

/**
 * A list of all the routes in the app.
 */
val routes = listOf(MainRouteSpec, CompassRouteSpec, SettingsRouteSpec)

/**
 * The default route to use when navigating.
 */
val defaultRoute = MainRouteSpec

File: apps/iphone/Bönetider/Bönetider/Routes/RouteSpec.swift
//
//  RouteSpec.swift
//  Bönetider
//
//  Created by Thanish Nizam on 2025-01-30.
//

import SwiftUI

// RouteSpec describes a route
protocol RouteSpec {
  var name: String { get }
  var text: String { get }
  var icon: String { get }
  var content: any View { get }
}

// AvailableRoutes is a list of all routes used by the application.
var AvailableRoutes: [RouteSpec] = [
  HomeRouteSpec(),
  CompassRouteSpec(),
  SettingsRouteSpec()
]

// DefaultRoute is the route that will be shown when the app is opened.
var DefaultRoute: RouteSpec {
  AvailableRoutes.first!
}

// ContentRoutes are routes with content ( does not include settings ).
var ContentRoutes: [RouteSpec] {
  AvailableRoutes.filter { $0.name != "settings" }
}
